subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaBasePlugin.java

```diff
@@ -305,20 +307,27 @@ private void defineConfigurationsForSourceSet(SourceSet sourceSet, Configuration
     }
 
     private void configureCompileDefaults(final Project project, final JavaPluginConvention javaConvention) {
+        JavaPluginExtension javaExtension = project.getExtensions().getByType(JavaPluginExtension.class);
+        DefaultJavaPluginConvention defaultJavaPluginConvention = (DefaultJavaPluginConvention) javaConvention;
         project.getTasks().withType(AbstractCompile.class).configureEach(compile -> {
             ConventionMapping conventionMapping = compile.getConventionMapping();
-            conventionMapping.map("sourceCompatibility", determineCompatibility(compile, javaConvention::getSourceCompatibility));
-            conventionMapping.map("targetCompatibility", determineCompatibility(compile, javaConvention::getTargetCompatibility));
+            conventionMapping.map("sourceCompatibility", determineCompatibility(compile, javaExtension, defaultJavaPluginConvention::getSourceCompatibility, defaultJavaPluginConvention::getRawSourceCompatibility));
+            conventionMapping.map("targetCompatibility", determineCompatibility(compile, javaExtension, defaultJavaPluginConvention::getTargetCompatibility, defaultJavaPluginConvention::getRawTargetCompatibility));
         });
     }
 
-    private Callable<Object> determineCompatibility(AbstractCompile compile, Supplier<JavaVersion> javaVersionSupplier) {
+    private Callable<Object> determineCompatibility(AbstractCompile compile, JavaPluginExtension javaExtension, Supplier<JavaVersion> javaVersionSupplier, Supplier<JavaVersion> rawJavaVersionSupplier) {
         return () -> {
             if (compile instanceof JavaCompile) {
                 JavaCompile javaCompile = (JavaCompile) compile;
                 if (javaCompile.getOptions().getRelease().isPresent()) {
+                    // Release set on the task wins, no need to check *Compat has having both is illegal anyway
                     return JavaVersion.toVersion(javaCompile.getOptions().getRelease().get()).toString();
                 } else if (javaCompile.getJavaCompiler().isPresent()) {
+                    // Toolchains in use
+                    if (((DefaultToolchainSpec) javaExtension.getToolchain()).isConfigured() && rawJavaVersionSupplier.get() != null) {
+                        throw new InvalidUserDataException("The new Java toolchain feature cannot be used at the project level in parallel with source and/or target compatibility");
+                    }
```

Comment:
consider pulling this into a method like `ensureToolchainDoesNotOverlapWithCompatibility` (or a better name :) )